llvm/lib/Transforms/Coroutines/CoroSplit.cpp

```diff
@@ -1203,10 +1203,6 @@ static bool simplifyTerminatorLeadingToRet(Instruction *InitialInst) {
       if (isa<BitCastInst>(I) || I->isDebugOrPseudoInst() ||
           I->isLifetimeStartOrEnd())
         I = I->getNextNode();
-      else if (isInstructionTriviallyDead(I))
-        // Duing we are in the middle of the transformation, we need to erase
-        // the dead instruction manually.
-        I = &*I->eraseFromParent();
```

Comment:
can we use `RecursivelyDeleteTriviallyDeadInstructions()` here instead? simplify seems overkill.

also do you know why we would end up with dead instructions at this point? we should have already simplified the function when we get to corosplit

Reply:
> can we use RecursivelyDeleteTriviallyDeadInstructions() here instead? simplify seems overkill.

In my test case the first instruction isn't trivially dead, so I don't think that would work.

> do you know why we would end up with dead instructions at this point? we should have already simplified the function when we get to corosplit

I think such instructions can emerge during corosplit. Maybe @ChuanqiXu9 can answer that one.

Reply:
> > can we use RecursivelyDeleteTriviallyDeadInstructions() here instead? simplify seems overkill.

> In my test case the first instruction isn't trivially dead, so I don't think that would work.

if we call it on the second instruction, it'll delete the first one as well

> > do you know why we would end up with dead instructions at this point? we should have already simplified the function when we get to corosplit

> I think such instructions can emerge during corosplit. Maybe @ChuanqiXu9 can answer that one.

understanding this a bit better would be good to see if we're doing extra unnecessary work, or if those instructions should be cleaned up earlier in the pass

Reply:
Yeah, since these instructions become dead during CoroSplit pass. The internals are

```
  %save2 = call token @llvm.coro.save(ptr null)
  call fastcc void @fakeresume1(ptr align 8 null)
  %suspend2 = call i8 @llvm.coro.suspend(token %save2, i1 false)

  ; These (non-trivially) dead instructions are in the way.
  %gep = getelementptr inbounds i64, ptr %alloc.var, i32 0
  %foo = ptrtoint ptr %gep to i64

  switch i8 %suspend2, label %exit [
    i8 0, label %await.ready
    i8 1, label %exit
  ]
await.ready:
  call void @consume(ptr %alloc.var)
  call void @llvm.lifetime.end.p0(i64 1, ptr %alloc.var)
  br label %exit
exit:
  call i1 @llvm.coro.end(ptr null, i1 false, token none)
  ret void
}
```

During the splitting,  the value `%suspend2` will become a constant in different splitted functions. So it makes sense if the previous passes can't catch such things.

Reply:
> understanding this a bit better would be good to see if we're doing extra unnecessary work, or if those instructions should be cleaned up earlier in the pass

+1 I'd like to understand this better too :)

In the original code I was debugging, the presplit coroutine has a return value:

```
169:                                              ; preds = %166, %163, %135, %68, %48
  %170 = getelementptr inbounds i8, ptr %17, i64 -16
  %171 = ptrtoint ptr %170 to i64
  %172 = call i1 @llvm.coro.end(ptr null, i1 false) #19                                                                                 
  ret i64 %171                     
}
```

corosplit clones the function into various versions, and the modifications of those clones are what causes dead code, makes some values constant, etc.

In my case, `CoroCloner::create()` will make the original return unreachable:

```
  switch (Shape.ABI) {                                                  
  // In these ABIs, the cloned functions always return 'void', and the
  // existing return sites are meaningless.  Note that for unique 
  // continuations, this includes the returns associated with suspends;
  // this is fine because we can't suspend twice.
  case coro::ABI::Switch:                                                
  case coro::ABI::RetconOnce:
    // Remove old returns.
    for (ReturnInst *Return : Returns)
      changeToUnreachable(Return);                                     
    break;
```

and `replaceCoroEnds()` will replace the `@llvm.coro.end` by `ret void`.

So perhaps CoroCloner is the one who should clean up these instructions.

On the other hand, it seems corosplit relies on later passes to clean up this stuff, which is maybe fine (especially since it also runs at `-O0`), except that it's a problem for `addMustTailToCoroResumes()`.

> simplify seems overkill

Yes, maybe it's too big a hammer. On the other hand, the ad-hoc dce and constant folding in `simplifyTerminatorLeadingToRet()` makes me a little nervous.


Instead of deleting those dead instructions, maybe we could just look past them. What the code is really trying to do is follow the control flow to a return instruction. I think it should be able to skip any side-effect free instructions on the way as long as the values aren't needed for a conditional branch or such. I'll push a version which does that. Let me know what you think.

Reply:
> So perhaps CoroCloner is the one who should clean up these instructions.

Generally this is not something we'd like to do for the middle end passes. However, this is (was) a defect in the clang/llvm coroutines implementations. Since the symmetric transfer is semantics about C++ and according to our design for modules, we have to make it in the middle end. So this is the why we have this... But probably we can never get rid of this if we (Clang/LLVM) don't want to be conforming to C++ anymore.

> Yes, maybe it's too big a hammer. On the other hand, the ad-hoc dce and constant folding in simplifyTerminatorLeadingToRet() makes me a little nervous.

So my line is, if the problem comes from a particular C++ programs and the corresponding LLVM passes applied, yes we need to solve it. However, if it comes from a manual constructed IR only, then possibly we shouldn't do that.